addons/crm/tests/test_crm_lead.py

```diff
@@ -1009,3 +1011,233 @@ def test_crm_lead_mail_tracking_duration_batch(self):
+                msg = set_won_get_message(new_lead, user1)
+                self.assertEqual(msg, 'Wow, that was fast. That deal didn’t stand a chance!')
+
+        with self.mock_datetime_and_now(jan3_12pm):
+            interval = (jan3_12pm - jan2).total_seconds() / 60
+            self._update_duration_tracking(tracking_array, interval)
+
+            with self.subTest('Three-day streak'):
+                msg = set_won_get_message(leads[8], user1)
+                self.assertEqual(msg, 'You\'re on a winning streak. 3 deals in 3 days, congrats!')
+
+        with self.mock_datetime_and_now(jan3_1pm):
+            interval = (jan3_1pm - jan3_12pm).total_seconds() / 60
+            self._update_duration_tracking(tracking_array, interval)
+            msg = set_won_get_message(new_lead2, user1)
+
+            with self.subTest('First stage to last stage'):
+                self.assertEqual(msg, 'No detours, no delays - from stage1 straight to the win! 🚀')
+
+            with self.subTest('Check that no message is returned if no "achievement" is reached'):
+                self.assertFalse(set_won_get_message(leads[9], user1))
+                self.assertFalse(set_won_get_message(leads[10], user2))
+
+        # test lead rainbowman messages (leads with expected revenues)
+
+            with self.subTest('Team record for the last 30 days'):
+                leads_er[0].expected_revenue = 800
+                msg = set_won_get_message(leads_er[0], user1)
+                self.assertEqual(msg, 'Boom! Team record for the past 30 days.')
+
+            with self.subTest('Personal record for the last 30 days'):
+                leads_er[1].expected_revenue = 600
+                msg = set_won_get_message(leads_er[1], user2)
+                self.assertEqual(msg, 'You just beat your personal record for the past 30 days.')
+
+        with self.mock_datetime_and_now(jan6):
+            interval = (jan6 - jan3_1pm).total_seconds() / 60
+            self._update_duration_tracking(tracking_array, interval)
+
+            with self.subTest('No record, no message'):
+                msg = set_won_get_message(leads_er[2], user1)
+                self.assertFalse(msg)
+                msg = set_won_get_message(leads_er[3], user2)
+                self.assertFalse(msg)
+
+        with self.mock_datetime_and_now(jan11):
+            with self.subTest('Team record for the last 7 days'):
+                leads_er[4].expected_revenue = 650
+                msg = set_won_get_message(leads_er[4], user1)
+                self.assertEqual(msg, 'Yeah! Best deal out of the last 7 days for the team.')
+
+            with self.subTest('Personal record for the last 7 days'):
+                leads_er[5].expected_revenue = 550
+                msg = set_won_get_message(leads_er[5], user2)
+                self.assertEqual(msg, 'You just beat your personal record for the past 7 days.')
+
+        with self.mock_datetime_and_now(march1):
```

Comment:
For this one we just check that "it works again" after a month, probably not necessary to retest every case, so it's fine as is.